service/history/replication/stream_receiver.go

```diff
@@ -211,16 +227,69 @@ func (r *StreamReceiverImpl) recvEventLoop() error {
+	lowPriorityWaterMarkInfo := r.lowPriorityTaskTracker.LowWatermark()
+	size := r.highPriorityTaskTracker.Size() + r.lowPriorityTaskTracker.Size()
+
+	// no task received yet
+	if r.receiverMode == ReceiverModeUnset || (highPriorityWaterMarkInfo == nil && lowPriorityWaterMarkInfo == nil) {
 		return 0, nil
 	}
+
+	var highPriorityWatermark, lowPriorityWatermark *replicationpb.ReplicationState
+	inclusiveLowWaterMark := int64(math.MaxInt64)
+	var inclusiveLowWaterMarkTime time.Time
+
+	switch r.receiverMode {
+	case ReceiverModeTieredStack:
+		if highPriorityWaterMarkInfo == nil || lowPriorityWaterMarkInfo == nil {
+			// This is to prevent the case: high priority task tracker received a batch of tasks, but low priority task tracker did not get any task yet.
+			// i.e. HighPriorityTracker received watermark 10 and LowPriorityTracker did not receive any tasks yet.
+			// we should avoid ack with {high: 10, low: nil}. If we do, sender will not able to correctly interpret the overall low watermark of the queue,
+			// because it is possible that low priority tracker might receive a batch of tasks with watermark 5 later.
+			// It is also true for the opposite case.
+			r.logger.Info("Tiered stack mode. Have to wait for both high and low priority tracker received at least one batch of tasks before acking.")
+			return 0, nil
+		}
+		highPriorityWatermark = &replicationpb.ReplicationState{
+			InclusiveLowWatermark:     highPriorityWaterMarkInfo.Watermark,
+			InclusiveLowWatermarkTime: timestamppb.New(highPriorityWaterMarkInfo.Timestamp),
+		}
+		lowPriorityWatermark = &replicationpb.ReplicationState{
+			InclusiveLowWatermark:     lowPriorityWaterMarkInfo.Watermark,
+			InclusiveLowWatermarkTime: timestamppb.New(lowPriorityWaterMarkInfo.Timestamp),
+		}
+		if inclusiveLowWaterMark > highPriorityWaterMarkInfo.Watermark {
+			inclusiveLowWaterMark = highPriorityWaterMarkInfo.Watermark
+			inclusiveLowWaterMarkTime = highPriorityWaterMarkInfo.Timestamp
+		}
+		if inclusiveLowWaterMark > lowPriorityWaterMarkInfo.Watermark {
+			inclusiveLowWaterMark = lowPriorityWaterMarkInfo.Watermark
+			inclusiveLowWaterMarkTime = lowPriorityWaterMarkInfo.Timestamp
+		}
+	case ReceiverModeSingleStack:
+		if highPriorityWaterMarkInfo == nil { // This should not happen, more for a safety check
+			return 0, NewStreamError("Single stack mode. High priority tracker does not have low watermark info", serviceerror.NewInternal("Invalid tracker state"))
+		}
+		if lowPriorityWatermark != nil {
+			return 0, NewStreamError("Single stack mode. Should not receive low priority task", serviceerror.NewInternal("Invalid tracker state"))
+		}
+
+		highPriorityWatermark = &replicationpb.ReplicationState{
+			InclusiveLowWatermark:     highPriorityWaterMarkInfo.Watermark,
+			InclusiveLowWatermarkTime: timestamppb.New(highPriorityWaterMarkInfo.Timestamp),
+		}
+		inclusiveLowWaterMark = highPriorityWaterMarkInfo.Watermark
```

Comment:
does it mean we cannot update the mode? If we change from tiered to single, some tasks may lose?

Reply:
We can update the mode. The switch on mode is controlled by sender. When sender is switched from tiered mode to single, it should disconnect the stream. When stream re-established, sender knows it is in single mode now, so it will send task from min(ACK-HighPriority, ACK-LowPriority). In this way, we guaranteed no task lose.